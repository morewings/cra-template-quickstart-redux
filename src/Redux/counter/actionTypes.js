export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
