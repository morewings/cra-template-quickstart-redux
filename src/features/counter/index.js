export {default as CounterReducer} from './CounterReducer';
export {default as useCountValue} from './selectors';
export {default as useIncrementCounter} from './useIncrementCounter';
