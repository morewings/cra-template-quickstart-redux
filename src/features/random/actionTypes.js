export const GET_RANDOM_NUMBER = 'GET_RANDOM_NUMBER';
