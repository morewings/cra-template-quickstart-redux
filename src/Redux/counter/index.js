import * as actionTypes from './actionTypes';

export {default as CounterReducer} from './CounterReducer';
export {actionTypes};
