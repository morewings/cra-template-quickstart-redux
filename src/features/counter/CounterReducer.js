import {INCREMENT_COUNTER} from './actionTypes';

const initialState = {
  value: 0,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case INCREMENT_COUNTER:
      return {...state, value: action.value};

    default:
      return state;
  }
};
