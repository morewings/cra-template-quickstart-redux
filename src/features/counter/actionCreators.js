import {useCallback} from 'react';
import {useDispatch} from 'react-redux';
import useCountValue from './selectors';
import {INCREMENT_COUNTER} from './actionTypes';

const useActions = () => {
  const dispatch = useDispatch();
  const count = useCountValue();
  const incrementCounter = useCallback(() => {
    dispatch({
      type: INCREMENT_COUNTER,
      value: count + 1,
    });
  }, [count, dispatch]);
  return {incrementCounter};
};

export default useActions;
