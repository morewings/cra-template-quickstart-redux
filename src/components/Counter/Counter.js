import React from 'react';
import {useCountValue, useActions} from '../../features/counter';
import classes from './Counter.module.css';

const Counter = () => {
  /**
   *  Get count value from Redux store. We defined selector
   *  (state => state.counter.value) inside counter feature folder,
   *  to make component global state agnostic
   */
  const count = useCountValue();

  /** Create incrementCounter action, using custom hook from feature */
  const {incrementCounter} = useActions();

  return (
    <div className={classes.counter}>
      <h2 className={classes.header}>Sync counter</h2>
      <button
        className={classes.button}
        type="button"
        onClick={incrementCounter}>
        Increment by one
      </button>
      <div>
        Total value: <strong>{count}</strong>
      </div>
    </div>
  );
};

export default Counter;
