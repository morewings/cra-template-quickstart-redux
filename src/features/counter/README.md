## Counter

Stores value in the Redux state and increments by 1.

## Selectors

### `useCountValue`

Returns count value from the store.

```javascript
import {useCountValue} from 'features/counter';

// Needs to be run from inside React component or other hook.
const count = useCountValue();
```

## Action creators

### `useIncrementCounter`

Increment stored count value by one.

```javascript
import {useIncrementCounter} from 'features/counter';

// Needs to be run from inside React component or other hook.
const incrementCounter = useIncrementCounter();
const handleClick = () => {
  useIncrementCounter();
}
```

