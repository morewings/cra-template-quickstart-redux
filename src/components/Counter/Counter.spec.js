import React from 'react';
import {Provider} from 'react-redux';
import {render, fireEvent} from '@testing-library/react';
import configureStore from 'redux-mock-store';
import {INCREMENT_COUNTER} from 'features/counter/actionTypes';
import Counter from './Counter';

describe('components > Counter', () => {
  /** Create mock store with the count value */
  const mockStore = configureStore([]);
  const store = mockStore({
    count: {
      value: 6,
    },
  });

  /**
   * Add spy to watch for store.dispatch method.
   * @see https://jestjs.io/docs/en/jest-object#jestspyonobject-methodname
   */
  jest.spyOn(store, 'dispatch');

  /**
   * Jest hook which runs before each test,
   * @see https://jestjs.io/docs/en/api#beforeeachfn-timeout
   */
  beforeEach(() => {
    /**
     * Clear any saved mock data from previous tests,
     * because jest saves calls data for spies and mocks.
     * @see https://jestjs.io/docs/en/mock-function-api#mockfnmockclear
     */
    store.dispatch.mockClear();
  });

  it('renders without crashing', () => {
    /**
     * `asFragment`:
     * @see https://testing-library.com/docs/react-testing-library/api#asfragment
     * `qetByText`:
     * @see https://testing-library.com/docs/dom-testing-library/api-queries#bytext
     * `wrapper`
     * @see https://testing-library.com/docs/react-testing-library/api#wrapper
     */
    const {asFragment, getByText} = render(<Counter />, {
      wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
    });

    /**
     * Basic snapshot test to make sure, that rendered component
     * matches expected footprint.
     */
    expect(asFragment()).toMatchSnapshot();

    /** More precise test for counter value */
    expect(getByText(/6/i).textContent).toBe('6'); // 6 is value we expect, we need to convert Number to String, because HTMLElement textContent method returns string value
  });

  it('dispatches an action on button click', () => {
    /**
     * `getByRole`:
     * @see https://testing-library.com/docs/dom-testing-library/api-queries#byrole
     */
    const {getByRole} = render(<Counter />, {
      wrapper: ({children}) => <Provider store={store}>{children}</Provider>,
    });

    /**
     * Search for the button and make testing library click on it
     * @see https://testing-library.com/docs/react-testing-library/cheatsheet#events
     */
    fireEvent.click(getByRole('button'));

    /** Check if store.dispatch method was run */
    expect(store.dispatch).toHaveBeenCalledTimes(1);

    /** Check if store.dispatch was run with correct action */
    expect(store.dispatch).toHaveBeenCalledWith({
      type: INCREMENT_COUNTER,
      value: 7,
    });
  });
});
