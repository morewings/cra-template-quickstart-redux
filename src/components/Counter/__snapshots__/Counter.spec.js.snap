// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Counter renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="counter"
  >
    <h2
      class="header"
    >
      Sync counter
    </h2>
    <button
      class="button"
      type="button"
    >
      Increment by one
    </button>
    <div>
      Total value: 
      <strong>
        6
      </strong>
    </div>
  </div>
</DocumentFragment>
`;
